### ISSUE TITLE
some problems during the SLT training and testing phase

### ISSUE THREAD

**Issue by honoryaoyao** [zhoubenjia/GFSLT-VLP#13]
Hello, I replaced FeatureExtracter in the SLT stage. After training for some time, its accuracy has not improved. I used the best checkpoint to do the test, and every prediction result was consistent. What is the reason for this?

train log
![train_log](https://github.com/zhoubenjia/GFSLT-VLP/assets/109032131/73361d1b-2bd4-4d9c-be1d-8f6599c76cf4)

**Comment by honoryaoyao** [comment]
test result
![tmp_pres](https://github.com/zhoubenjia/GFSLT-VLP/assets/109032131/819e0369-5cae-4815-8802-bf2ceedc8840)

**Comment by zhoubenjia** [comment]
> Hello, I replaced FeatureExtracter in the SLT stage. After training for some time, its accuracy has not improved. I used the best checkpoint to do the test, and every prediction result was consistent. What is the reason for this?
> 
> train log ![train_log](https://private-user-images.githubusercontent.com/109032131/325458858-73361d1b-2bd4-4d9c-be1d-8f6599c76cf4.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwNTA0MTIsIm5iZiI6MTcxNDA1MDExMiwicGF0aCI6Ii8xMDkwMzIxMzEvMzI1NDU4ODU4LTczMzYxZDFiLTJiZDQtNGQ5Yy1iZTFkLThmNjU5OWM3NmNmNC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNVQxMzAxNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMGZkMGIzODZhMjEzMDdkZmYwYWY5NjY2YjIwY2ZhNDA0Y2FlZGRlZDNkNzQwMDcxNGRiMjI4NWY1OWI1ODlkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.hratp6Bez7vSr53olaRwUtb7TVPpWy1tF2GxmBGVrhA)

Certainly:

Could you share some of your code please?  The FeatureExtractor() function typically comprises a spatial feature extraction module and a local time modeling module. Both components are crucial for convergence. My suspicion is that you might have overlooked the latter, or perhaps omitted the normalization operation (BN & LN) on the features.

**Comment by honoryaoyao** [comment]
Thank you very much. At present, I use Vmamba as the new visual extraction backbone, and the other parts are using the modules provided in your code. The following is the code of model.

class SignModel(nn.Module):
    def __init__(self, args, embed_dim=1024):
        super(SignModel, self).__init__()
        self.args = args
        self.fc = nn.Linear(768, 1024)
        # pht-event
        self.mbart = MBartForConditionalGeneration.from_pretrained("/media/amax/c08a625b-023d-436f-b33e-9652dc1bc7c0/DATA/rongyao/newcode/benchmark/GFSLT-VLP-main/pretrain_models/mytran")
        # our_sl
        # self.mbart = MBartForConditionalGeneration.from_pretrained("/media/amax/c08a625b-023d-436f-b33e-9652dc1bc7c0/DATA/rongyao/newcode/GFSLT-VLP-main/pretrain_models_new/our_sl_mbart_char_ry")
        
        self.conv_1d = TemporalConv(input_size=768, hidden_size=1024, conv_type=2)
        self.sign_emb = V_encoder(emb_size=embed_dim, feature_size=embed_dim)
        self.embed_scale = 1.0
        
        # Vmamba
        self.visual_model = create_model(type_vssm='tiny')
        load_pretrain(self.visual_model, ckpt="/media/amax/c08a625b-023d-436f-b33e-9652dc1bc7c0/DATA/rongyao/newcode/R2Gen-videomamba-224/vssm_tiny_0230_ckpt_epoch_262.pth")
        #计算模型的参数量
        visual_params = sum(p.numel() for p in self.visual_model.parameters())
        mbart_params = sum(p.numel() for p in self.mbart.parameters())
        print("visual_model parameters:", visual_params)
        print("mbart parameters:", mbart_params)
       

    def __str__(self):
        model_parameters = filter(lambda p: p.requires_grad, self.parameters())
        params = sum([np.prod(p.size()) for p in model_parameters])
        return super().__str__() + '\nTrainable parameters: {}'.format(params)
    
    def visual_forward(self, src_input):
        images = src_input['input_ids'].cuda()
        attention_mask = src_input['attention_mask']

        out_visual = self.visual_model(images)
        BT, C_ = out_visual.shape
        B = src_input['attention_mask'].shape[0]
        re_visual = out_visual.reshape(B, -1, C_)
        
        frames_feature = self.conv_1d(re_visual)
        
        inputs_embeds = self.sign_emb(frames_feature) 
        inputs_embeds = self.embed_scale * inputs_embeds 

        
        return inputs_embeds, attention_mask

    def forward(self, src_input, tgt_input):
      
        inputs_embeds, attention_mask = self.visual_forward(src_input)
        
        out = self.mbart(inputs_embeds = inputs_embeds,
                    attention_mask = attention_mask.cuda(),
                    # decoder_input_ids = tgt_input['input_ids'].cuda(),
                    labels = tgt_input['input_ids'].cuda(),
                    decoder_attention_mask = tgt_input['attention_mask'].cuda(),
                    return_dict = True,
                    )

        return out['logits']

    def generate(self,src_input, max_new_tokens, num_beams, decoder_start_token_id):
        
        inputs_embeds, attention_mask = self.visual_forward(src_input)
        out = self.mbart.generate(inputs_embeds = inputs_embeds,
                    attention_mask = attention_mask.cuda(),max_new_tokens=max_new_tokens,num_beams = num_beams,
                                decoder_start_token_id=decoder_start_token_id
                            )
        return out

**Comment by honoryaoyao** [comment]
Thank you very much. At present, I use Vmamba as the new visual extraction backbone, and the other parts are using the modules provided in your code. The following is the code of model.


```
class SignModel(nn.Module):
    def __init__(self, args, embed_dim=1024):
        super(SignModel, self).__init__()
        self.args = args

        self.mbart = MBartForConditionalGeneration.from_pretrained("/media/amax/c08a625b-023d-436f-b33e-9652dc1bc7c0/DATA/rongyao/newcode/benchmark/GFSLT-VLP-main/pretrain_models/mytran")

        self.conv_1d = TemporalConv(input_size=768, hidden_size=1024, conv_type=2)
        self.sign_emb = V_encoder(emb_size=embed_dim, feature_size=embed_dim)
        self.embed_scale = 1.0
        
        self.visual_model = create_model(type_vssm='tiny')
        load_pretrain(self.visual_model, ckpt="/media/amax/c08a625b-023d-436f-b33e-9652dc1bc7c0/DATA/rongyao/newcode/R2Gen-videomamba-224/vssm_tiny_0230_ckpt_epoch_262.pth")
       
        visual_params = sum(p.numel() for p in self.visual_model.parameters())
        mbart_params = sum(p.numel() for p in self.mbart.parameters())
        print("visual_model parameters:", visual_params)
        print("mbart parameters:", mbart_params)
       

    def __str__(self):
        model_parameters = filter(lambda p: p.requires_grad, self.parameters())
        params = sum([np.prod(p.size()) for p in model_parameters])
        return super().__str__() + '\nTrainable parameters: {}'.format(params)
    
    def visual_forward(self, src_input):
        images = src_input['input_ids'].cuda()
        attention_mask = src_input['attention_mask']

        out_visual = self.visual_model(images)
        BT, C_ = out_visual.shape
        B = src_input['attention_mask'].shape[0]
        re_visual = out_visual.reshape(B, -1, C_)
        
        frames_feature = self.conv_1d(re_visual)
        
        inputs_embeds = self.sign_emb(frames_feature) 
        inputs_embeds = self.embed_scale * inputs_embeds 

        
        return inputs_embeds, attention_mask

    def forward(self, src_input, tgt_input):
      
        inputs_embeds, attention_mask = self.visual_forward(src_input)
        
        out = self.mbart(inputs_embeds = inputs_embeds,
                    attention_mask = attention_mask.cuda(),
                    # decoder_input_ids = tgt_input['input_ids'].cuda(),
                    labels = tgt_input['input_ids'].cuda(),
                    decoder_attention_mask = tgt_input['attention_mask'].cuda(),
                    return_dict = True,
                    )

        return out['logits']

    def generate(self,src_input, max_new_tokens, num_beams, decoder_start_token_id):
        
        inputs_embeds, attention_mask = self.visual_forward(src_input)
        out = self.mbart.generate(inputs_embeds = inputs_embeds,
                    attention_mask = attention_mask.cuda(),max_new_tokens=max_new_tokens,num_beams = num_beams,
                                decoder_start_token_id=decoder_start_token_id
                            )
        return out
```

**Comment by honoryaoyao** [comment]
Hello, thank you very much for your reply. Is it convenient for you? We can use wechat to communicate. My wechat ID is ry277550138.
